### PR TITLE
Remove dependency on pipe events in HttpConnection

### PR DESCRIPTION
**Pull request by davidfowl**
- Refactored the HttpConnection to not depend on OnReaderCompleted and OnWriterCompleted. Instead we use ConnectionClosed to detect the FIN that we propagate via ConnectedAborted.
- Fire ConnectionClosed when a FIN is received from both transports.
- Remove pipe completion from Http1Connection and Http1OutputProducer. Instead just return from request processing.
- Flow the CancellaitonToken property in RawStream

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs
Comment (gfoidl) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]:
```suggestion
                error = ex as ConnectionAbortedException;
```

**Comment by jkotalik** [comment]
I'd also prefer to have this for https://github.com/aspnet/AspNetCore/pull/11109

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs]: 👏

**Review by jkotalik — APPROVED** [review]
The change itself seems good, but I think we should go over this in person before merging. Also to figure out how we can sync this with https://github.com/aspnet/AspNetCore/pull/11109/checks?check_run_id=146487478. I'd rather merge this first before merging my PR as I don't want to have the pipe scheduler be thread pool.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: But the perf! JK. looking forward to the update.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: Shouldn't we wait for the connection middleware pipeline before closing the input? I think this breaks half-duplex if that matters.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: We need a flag to know *we're* the once that canceled the read.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: Why not make this an empty `catch (ConnectionAbortedException) { }` block?

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: I misread. You are only storing this if it *is* a ConnectionAbortedException. In that case having a non-empty ConnectionAbortedException catch then an empty OperationCanceledException catch after is still probably better.

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: This is less code.

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: This makes me wish we called to ConnectionClosed token FinReceived or something like that, because the connection might be only half closed at this point.

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: I see why you added this bool for the libuv transport, but it seems completely unnecessary in the socket transport. FireConnectionClosed can only ever be called once.

File: src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: I see you're taking advantage of the single-threaded nature of libuv's event loop 👍

File: src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: If we're going to continue firing ConnectionClosed for libuv output errors, we should do the same for the socket transport.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs]: Why is this necessary?

File: src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: Now that we're flowing the cancellationToken through RawStream, do we have a regression test?

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: Yes that is true. I noticed that when I was copying code.

File: src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: What do you mean? Aren't we doing that? It always fires no?

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs
Comment (halter73) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs]: Discussed this offline. Calling ReceiveEnd unblocks the ProcessRequest loop allowing to observe the FIN and write a 4XX before the transport is aborted.

File: src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs
Comment (halter73) [reply src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: I'm now thinking about making this more like the socket transport instead. The important think is consistency.

Instead of calling FireConnectionClosed here, just don't dispose _socket here. Instead, close the uv_stream_handle only after both the output loop completed and

a) you've received a terminal read callback or
b) the connection was aborted

File: src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: Why not fire this for any terminal status?

Comment (halter73) [reply src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs]: Really we should be calling this every time we call Input.Complete().

File: src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: Done

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: You could argue the same about HttpContext.ConnectionAborted, I don't think it makes a difference.

**Comment by davidfowl** [comment]
@aspnet/build What's the deal with all the PRs, do we merge, do we wait? I know it's being looked at but what is the mitigation to unblock work?

File: src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: Does CancelPendingRead need to be synchronized?

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: Not for our usage.

**Comment by dougbu** [comment]
@davidfowl as a few people have done, I suggest making a polite request for an admin to merge PRs where build success isn't being reported correctly or individual jobs hang. Looks like both of those issues have hit this PR but is it ready to merge?

**Comment by davidfowl** [comment]
> @davidfowl as a few people have done, I suggest making a polite request for an admin to merge PRs where build success isn't being reported correctly or individual jobs hang. Looks like both of those issues have hit this PR but is it ready to merge?

I can merge I just want to know the protocol. Yes this is ready to merge.

**Comment by dougbu** [comment]
:ok:, ping if you're blocked

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs
Comment (halter73) [reply src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: You absolutely could, but at least that's HTTP-specific so it makes a bit more sense.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs]: I don't even think _cancelCalled *needs* to be volatile since CancelPendingRead acquires and releases the Pipe's `_sync` lock right afterwards.

**Pull request merged**
